---
title: "Riboviz32test"
author: "FlicAnderson"
date: "27/03/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(getopt, quietly=T)
library(here)
suppressMessages(library(Rsamtools))
suppressMessages(library(rtracklayer))
suppressMessages(library(rhdf5))
suppressMessages(library(parallel))
suppressMessages(library(optparse))
suppressMessages(library(RcppRoll))
suppressMessages(library(ggplot2))
suppressMessages(library(tidyr))
suppressMessages(library(dplyr))
suppressMessages(library(magrittr))
suppressMessages(library(purrr))
suppressMessages(library(here))

  # this_script <- "generate_stats_figs.R"
  # path_to_this_script <- here("rscripts", this_script)
  # source(here("rscripts", "provenance.R"))
  # source(here("rscripts", "read_count_functions.R"))
```


```{r cars, eval=TRUE}

# NOTE: other libraries loaded from read_count_functions.R  

# Handle interactive session behaviours or use get_Rscript_filename():
if (interactive()) { print("yes")
  # # Use hard-coded script name and assume script is in "rscripts"
  # # directory. This assumes that interactive R is being run within
  # # the parent of rscripts/ but imposes no other constraints on
  # # where rscripts/ or its parents are located.
  this_script <- "generate_stats_figs.R"
  path_to_this_script <- here::here("rscripts", this_script)
  source(here::here("rscripts", "provenance.R"))
  source(here::here("rscripts", "read_count_functions.R"))
} else { print("no")
  # # Deduce file name and path using reflection as before.
  # this_script <- getopt::get_Rscript_filename()
  # path_to_this_script <- this_script
  # source(file.path(dirname(this_script), "provenance.R")) # FLIC: consider if this can be redone using here() too
  # source(file.path(dirname(this_script), "read_count_functions.R")) # FLIC: consider if this can be redone using here() too
  this_script <- "generate_stats_figs.R"
  path_to_this_script <- here::here("rscripts", this_script)
  source(here::here("rscripts", "provenance.R"))
  source(here::here("rscripts", "read_count_functions.R"))
}

```

```{r cars1}

# define input options for optparse package
option_list <- list(
  make_option("--output-dir",
              type = "character", default = "./",
              help = "Output directory"
  ),
  make_option("--orf-fasta-file",
              type = "character", default = FALSE,
              help = "FASTA file with nt seq"
  ),
  make_option("--orf-gff-file",
              type = "character", default = NA,
              help = "riboviz generated GFF2/GFF3 annotation file"
  ),
  make_option("--num-processes",
              type = "integer", default = 1,
              help = "Number of cores for parallelization"
  ),
  make_option("--min-read-length",
              type = "integer", default = 10,
              help = "Minimum read length in H5 output"
  ),
  make_option("--max-read-length",
              type = "integer", default = 50,
              help = "Maximum read length in H5 output"
  ),
  make_option("--buffer",
              type = "integer", default = 250,
              help = "Length of flanking region around the CDS"
  ),
  make_option("--primary-id",
              type = "character", default = "gene_id",
              help = "Primary gene IDs to access the data (YAL001C, YAL003W, etc.)"
  ),
  make_option("--dataset",
              type = "character", default = "vignette",
              help = "Name of the dataset"
  ),
  make_option("--rpf",
              type = "logical", default = TRUE,
              help = "Is the dataset an RPF or mRNA dataset?"
  ),
  make_option("--features-file",
              type = "character", default = NA,
              help = "features file, columns are gene features and rows are genes"
  ),
  make_option("--do-pos-sp-nt-freq",
              type = "logical", default = TRUE,
              help = "do calculate the position-specific nucleotide frequency"
  ),
  make_option("--t-rna-file",
              type = "character", default = NA,
              help = "tRNA estimates in .tsv file"
  ),
  make_option("--codon-positions-file",
              type = "character", default = NA,
              help = "Codon positions in each gene in .Rdata file"
  ),
  make_option("--count-threshold",
              type = "integer", default = 64,
              help = "threshold for count of reads per gene to be included in plot"
  ),
  make_option("--length_threshold",
              type = "integer", default = 500,
              help = "threshold for length of genes to be included in metagene plots"
  ),
  make_option("--output-prefix",
              type = "character", default = "",
              help = "Prefix for output files"
  ),
  make_option("--hd-file",
              type = "character", default = "output.h5",
              help = "Location of H5 output file"
  ),
  make_option("--nnt-buffer",
              type = "integer", default = 25,
              help = "n nucleotides of UTR buffer to include in metagene plots"
  ),
  make_option("--nnt-gene",
              type = "integer", default = 50,
              help = "n nucleotides of gene to include in metagene plots"
  ),
  make_option("--asite-disp-length-file",
              type = "character", default = NA,
              help = "asite displacement file
    table with one displacement per read length"
  )
)

```


```{r cars2}
# print provenance
print_provenance(path_to_this_script)
```

```{r cars3}
# read in commandline arguments
opt <- optparse::parse_args(OptionParser(option_list = option_list),
                            convert_hyphens_to_underscores=TRUE)
```

```{r cars4}
#### VIGNETTE 1 SAMPLE

opt$output_dir <- "vignette/output/WTnone"
opt$orf_fasta_file <- "vignette/input/yeast_YAL_CDS_w_250utrs.fa"
opt$orf_gff_file <- "vignette/input/yeast_YAL_CDS_w_250utrs.gff3" 
opt$num_processes <- 8
opt$min_read_length <- 10
opt$max_read_length <- 50
opt$buffer <- 250
opt$primary_id <- "Name"
opt$dataset <- "vignette"
opt$rpf <- TRUE
opt$features_file <- "data/yeast_features.tsv"
opt$do_pos_sp_nt_freq <- TRUE
opt$t_rna_file <- "data/yeast_tRNAs.tsv"
opt$codon_positions_file <- "data/yeast_codon_pos_i200.RData"
opt$count_threshold <- 64
opt$hd_file <- "vignette/output/WTnone/WTnone.h5"
opt$asite_disp_length_file <- "vignette/input/asite_disp_length_yeast_standard.txt"

# attach opt list to be able to refer to variables in the list by names alone
# ie `height` rather than `women$height`
attach(opt)
```

```{r cars5}
print("generate_stats_figs.R running with parameters:")
opt
```

```{r cars6}
# prepare files, opens hdf5 file connection
#hdf5file <- rhdf5::H5Fopen(hd_file) # filehandle for the h5 file
rhdf5::h5closeAll()
```


```{r cars7}
head(h5ls(here(opt$hd_file), recursive = FALSE))
#str(a)

# list of gene names taken from h5 file
gene_names <- rhdf5::h5ls(here::here(opt$hd_file), recursive = 1)$name

head(gene_names)

```

```{r cars8}
# read in coding sequences
print(orf_fasta_file)

coding_seqs <- readDNAStringSet(here(orf_fasta_file)) 
```
Have to wrap `orf_fasta_file` in here() as the path isn't full path otherwise. here() fixes this nicely and runs 

```{r cars9}
# range of read lengths between parameters set in config file
read_range <- min_read_length:max_read_length
```

```{r cars10}
# read in positions of all exons/genes in GFF format and convert to tibble data frame 
print(orf_gff_file)
gff_df <- readGFFAsDf(here(orf_gff_file))
```

```{r cars11}
# set ggplot2 theme for plots drawn after this; use dark on light theme
ggplot2::theme_set(theme_bw())
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
